### ISSUE TITLE
x64_dbg tries to call CBWINEVENTGLOBAL even after plugin was unloaded, which can cause a crash

### ISSUE THREAD

**Issue by Nukem9** [x64dbg/x64dbg#191]
Based on this report, which I can reproduce only on Windows XP:

https://forum.tuts4you.com/topic/21128-multiline-ultimate-assembler/?p=167305

As I have already proposed in a mail conversation to mrexodia, I suggest to unload plugins only after all debugger windows were destroyed and message loop was exited, possibly right before process exit. This will lead to a cleaner design, and at the same time will fix this, and other bugs.

Also I suggest to add a `plugdestroy` callback function in addition to the `plugstop` function. The `plugstop` function will be called right after the choice to close the debugger was made, and `plugdestroy` will be called right before the process exits.

So basically, something like this:

1. User clicks on "Exit".
2. Debugger calls `plugstop`.
3. Debugger closes the main window, finalizes the GUI, exits the message loop.
4. Debugger calls `plugdestroy`.
5. Debugger process ends.

----------------------------------------
- Bitbucket: https://bitbucket.org/mrexodia/x64_dbg/issue/191
- Originally reported by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)
- Originally created at: 2014-09-04T22:43:55.207

**Comment by Nukem9** [comment]
This is already what happens right now. The reason the thing crashes is that I didn't clear the plugin list, which basically means it will try to call plugins that are unloaded using FreeLibrary.

There will not be any `plugindestroy` function. It is 100% useless, except for the fact that plugins can add a dialog to ask if users want to quit. If you want to do operations on target stop (like exit stack analysis or something), use the CB_DEBUG* functions. If you need to complete something, do it in the plugstop function. The debugger will wait for your stuff to be completed.

This was fixed in bf8c69b475b94417dfb2a86cd761c82a202bc2c6. Right now, when I leave the MultiASM window open and close the debugger it will simply close the MultiASM window and exit without crashes.

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
> when I leave the MultiASM window open and close the debugger it will simply close the MultiASM window and exit without crashes.

I don't see how the commit you've linked fixes this issue. After all, the window procedure of multiasm is not called by x64_dbg, but by the system, so freeing the library right after calling `plugstop` still can cause a crash.

Also, I believe that introducing the `plugindestroy` callback function will lead to a cleaner, more correct API design. But in the end, it's your debugger and your choice. All I can say is that I'm reluctant to search for workarounds for crashes which are caused by this, while it works perfectly well with OllyDbg.

----------------------------------------
Original comment by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)

**Comment by Nukem9** [comment]
The problem (which was an actual bug in x64_dbg) that was fixed by bf8c69b475b94417dfb2a86cd761c82a202bc2c6 had to do with when a global event message was sent to the event filter and then through `CB_WINEVENTGLOBAL`, the plugin callback processor would attempt to call the registered callback, even after `plugstop` was called, because the plugin list was not cleared. The commit ensured that the plugin list is cleared after `plugstop`, which means the `CB_WINEVENTGLOBAL` message would not be sent to freed plugins.

In your plugin you call `PostMessage(WM_CLOSE);`, which will not wait until the message was successfully processed. This means its a pure guess game if your window is closed before you return from plugstop. When the `WM_CLOSE` message is still being processed and at that moment the DLL is freed, it will generally crash. However, when you call `SendMesssage(WM_CLOSE);` instead, it will only return **after** you cleaned up your window, meaning you don't have to worry about random crashes.

Probably the only reason it works correctly in OllyDbg is that there is some delay between `ODBG_Pluginclose` and `ODBG_Plugindestroy`.

The answer to [this](http://stackoverflow.com/questions/3376619/what-is-the-difference-between-send-message-and-post-message-and-how-these-relat) question is interesting regarding this matter:
>PostMessage (in "pure windows programming", aka win32 API) is asynchronous, i.e., to quote the docs:

>>Places (posts) a message in the message queue associated with the thread that created the specified window and returns without waiting for the thread to process the message.
>
>>To post a message in the message queue associated with a thread, use the PostThreadMessage function.```
>
>SendMessage is synchronous, that is, again quoting:
>
>>Sends the specified message to a window or windows. The SendMessage function calls the window procedure for the specified window and does not return until the window procedure has processed the message.`

>>To send a message and return immediately, use the SendMessageCallback or SendNotifyMessage function. To post a message to a thread's message queue and return immediately, use the PostMessage or PostThreadMessage function.

Like I said before, adding a `plugdestroy` message would just be:
```
bool _dbg_exitsignal()
{
    if(!plugstop())
        return false; //do not exit
    //some other cleanup code
    plugdestroy();
    return true; //exit
}
```

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
> The problem ... a global event message was sent to the event filter and then through CB_WINEVENTGLOBAL ...

Yeah, this bug was probably fixed by that commit, but not the bug of closing the debugger when the multiasm window is open.

> In your plugin you call PostMessage(WM_CLOSE);, which will not wait ...

Sending `WM_CLOSE` won't wait either. I've just rechecked to make sure, and even if I use `SendMesssage(WM_CLOSE);`, the window procedure receives calls after `SendMesssage` returns.

> Probably the only reason it works correctly in OllyDbg is that there is some delay between ODBG_Pluginclose and ODBG_Plugindestroy.

No, there's no delay and no luck involved. I've tried to explain many times how OllyDbg manages this, but you don't seem to understand.

`ODBG_Pluginclose` is called early, right after you click on close. OllyDbg's main window still exists and shown, nothing was cleaned up, etc.

`ODBG_Plugindestroy` is called lately, after all GUI is finalized! The message loop is finalized as well, ALL the windows of the process are gone, so multiasm's windows procedure just CANNOT be called anymore. All is left for the process is to exit. You might want to look again at the five stages I written in the original bug report above.

> The answer to this question is interesting regarding this matter:

Yes, but `WM_CLOSE` doesn't destroy the window right away. It just schedules a closing event.

> Like I said before, adding a plugdestroy message would just be:

It shouldn't be like that! That obviously won't help!

It should be more like this:

```
void main()
{
    // ...

    run_GUI(); // <-- on_exit_button_click() gets called here

    // ...

    before_termination();
    exit_process();
}

bool on_exit_button_click()
{
    for(i=0; i<plugin_count; i++)
        if(!plugstop(i))
            return false; //do not exit
    return true; //exit
}

void before_termination()
{
    for(i=0; i<plugin_count; i++)
        plugdestroy(i);
}
```

----------------------------------------
Original comment by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)

**Comment by Nukem9** [comment]
ok, here is a small overview of what happens when I click the close button on the x64dbg window: 

```
#!c++

void MainWindow::closeEvent(QCloseEvent* event)
{
    hide(); //hide main window from sight (not the same as deleting)
    mCloseDialog->show(); //display 'Closing debugger...' dialog
    mCloseDialog->setFocus(); //set focus to this dialog
    CloseHandle(CreateThread(0, 0, closeThread, this, 0, 0)); //create closeThread
    if(bClose) //bClose = false on dialog init
        event->accept(); //accept event (close/destroy the main window)
    else
        event->ignore(); //ignore event (eg actually close the dialog) when bClose = false
}
```

After the closeEvent, x64_dbg will spawn a thread that does the following:

```
#!c++
DWORD WINAPI MainWindow::closeThread(void* ptr)
{
    static bool closing = false;
    if(closing) 
        return 0;
    closing = true; //ensure the thread doesn't run twice
    DbgExit(); //call exit of the Dbg
    MainWindow* mainWindow = (MainWindow*)ptr; //get pointer to MainWindow class
    mainWindow->bClose = true; //allow the function above this to actually close the dialog
    mainWindow->close(); //call the closeEvent
    return 0;
}
```

the DbgExit function looks (globally) like this:

```
#!c++

extern "C" DLL_EXPORT void _dbg_dbgexitsignal()
{
    cbStopDebug(0, 0); //stop debugging
    scriptabort(); //abort running scripts
    wait(WAITID_STOP); //after this, debugging stopped (wait for debugger to report it stopped)
    pluginunload(); //unload plugins
    TerminateThread(hCommandLoopThread, 0);
    CloseHandle(hCommandLoopThread);
    cmdfree(command_list);
    varfree();
    msgfreestack(gMsgStack);
    if(memleaks())
    {
        char msg[256] = "";
        sprintf(msg, "%d memory leak(s) found!\n\nPlease send 'alloctrace.txt' to the authors of x64_dbg.", memleaks());
        MessageBoxA(0, msg, "error", MB_ICONERROR | MB_SYSTEMMODAL);
    }
    else
        DeleteFileA(alloctrace);
    CriticalSectionDeleteLocks();
}
```

pluginunload looks like so:

```
#!c++

void pluginunload()
{
    int pluginCount = (int)pluginList.size();
    for(int i = pluginCount - 1; i > -1; i--)
    {
        PLUGSTOP stop = pluginList.at(i).plugstop;
        if(stop)
            stop(); //call plugstop
        plugincmdunregisterall(pluginList.at(i).initStruct.pluginHandle); //unregister callbacks
        FreeLibrary(pluginList.at(i).hPlugin); //actually unload the library
        pluginList.erase(pluginList.begin() + i); //erase plugin from list to prevent calling callbacks in functions that do not exist
    }
    pluginCallbackList.clear(); //remove all callbacks
    pluginMenuList.clear(); //clear menu list
    GuiMenuClear(GUI_PLUGIN_MENU); //clear the plugin menu
}
```

It already works like you propose, except that you clean up your mess before the debugger is completely terminated (the GUI is still there while cleaning up though).

Basically all that happens after the main window closed is:

```
#!c++
mConfiguration->save(); //save config on exit
return result; //return result from MainWindow
```

Event filters etc are **not** called after the close button was clicked, I just verified that. So I don't see where the bug is...

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
Hmm, looks messy (and not too thread-safe). Why do you need an extra thread?

`return result from MainWindow` - who receives it, and why can't you call `plugindestroy` there?

> Event filters etc are not called after the close button was clicked, I just verified that. So I don't see where the bug is...

The bug is not with event filters, but with the dialog's window procedure, which is called by the system.

----------------------------------------
Original comment by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)

**Comment by Nukem9** [comment]
fixed the thread-safety, although in practice you cannot click the close button fast enough to break it.

nobody receives that, it's just the return from main(). After the MainWindow was initialized (eg it has a HWND) it will call some initialization functions in the DBG (eg load plugins) and then it will start the application (MainWindow). plugstop gets called before MainWindow is destroyed (try calling `GetWindowText(GuiGetWindowHandle())`) so everything should be fine window-wise.

I also saw some interesting stuff about closing your dialog: http://msdn.microsoft.com/en-us/library/windows/desktop/ff381396(v=vs.85).aspx

Maybe you could try calling `DestroyWindow` directly or try calling `PostQuitMessage(0);` inside WM_DESTROY, this ends the message loop ([here](http://msdn.microsoft.com/en-us/library/windows/desktop/ff381396(v=vs.85).aspx) they talk about as if that is mandatory).

Unfortunately I cannot make the debugger crash on my system (not even with the video White uploaded at tuts4you). Do you have a good way of reproducing the crash (in the first post you say something about WinXP).


----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
> it's just the return from main()

So is this the main thread? Does it wait for the `closeThread` to finish? If so, this looks like a perfect place for `plugindestroy`.

> Maybe you could try calling DestroyWindow directly

That complicates stuff, as there might be child dialogs, such as the find dialog (Ctrl+F), which shouldn't be destroyed directly.

> or try calling PostQuitMessage(0); inside WM_DESTROY, this ends the message loop

This is 100% wrong. It ends the message loop of the thread, which is managed by Qt, not by multiasm.

> in the first post you say something about WinXP

The bug with Windows XP, the one White reported, was about the event filter callback and was probably fixed. I don't have a way to reproduce the unfixed bug, which is about timing.

----------------------------------------
Original comment by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)

**Comment by Nukem9** [comment]
the issue has been confirmed, needs fixing. The idea is to free the plugins somewhere before program termination, without adding any extra callbacks (extra callbacks = more complex plugin architecture)

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
oh, reproducing this on the dev build is (/was) very easy for me:

1. install multiasm in the `/plugins` dir
2. start x64_dbg
3. open the multiasm window
4. close x64_dbg
5. crash (sometimes in a random stack, sometimes with the correct stack)

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
> >  In your plugin you call PostMessage(WM_CLOSE);, which will not wait ...
> 
> Sending WM_CLOSE won't wait either. I've just rechecked to make sure, and even if I use SendMesssage(WM_CLOSE);, the window procedure receives calls after SendMesssage returns.

In fact, I might have been wrong here. I've changed it to `SendMessage`, changed a couple of other stuff, and looks like the call is synchronous now.

Please try [this version](https://www.dropbox.com/s/j4wddkvnefifvp0/multiasm_x64dbg_test.rar?dl=1) and tell me whether it's still crashing.

----------------------------------------
Original comment by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)

**Comment by Nukem9** [comment]
Seems to work fine without crashing now.

----------------------------------------
Original comment by: [Nukem9](http://bitbucket.org/Nukem9)

**Comment by Nukem9** [comment]
Okay, so can we call this issue resolved?

To be safe I might still implement an internal plugindestroy mechanism, but if the bug is solved that's just to prevent crashes like this from happening.

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)

**Comment by Nukem9** [comment]
I haven't tested this thoroughly, but if you are sure that the crash is gone, I guess that it can be closed.

----------------------------------------
Original comment by: [RaMMicHaeL](http://bitbucket.org/RaMMicHaeL)

**Comment by Nukem9** [comment]
just tested it, issue appears to be resolved :)

Thanks everyone!

----------------------------------------
Original comment by: [mrexodia](http://bitbucket.org/mrexodia)
